metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.lookup_advanced(string_kmer, check_reverse_complement);
+    }
+
+    sshash::lookup_result lookup_advanced_uint(
+            kmer_t uint_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.lookup_advanced_uint(uint_kmer, check_reverse_complement);
+    }
+
+    sshash::neighbourhood<kmer_t> kmer_forward_neighbours(
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.kmer_forward_neighbours(string_kmer, check_reverse_complement);
+    }
+
+    sshash::neighbourhood<kmer_t> kmer_backward_neighbours(
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.kmer_backward_neighbours(string_kmer, check_reverse_complement);
+    }
+
+  private:
+    // TODO: this is a hack since sshash::dictionary<kmer_t>::visit is always non-const
+    mutable sshash::dictionary<kmer_t> dict_;
+};
+
+constexpr DeBruijnGraph::node_index sshash_to_graph_index(uint64_t idx) {
+    return idx + 1;
+}
+constexpr uint64_t graph_index_to_sshash(DeBruijnGraph::node_index idx) {
+    return idx - 1;
+}
+
+DBGSSHash::DBGSSHash(size_t k, Mode mode) : k_(k), num_nodes_(0), mode_(mode) {
+    size_t odd_k = k_ % 2 == 0 ? k_ + 1 : k_;
+
+    if (odd_k * IDictionary::bits_per_char <= 64) {
+        dict_ = std::make_unique<Dictionary<Kmer64>>();
+    } else if (odd_k * IDictionary::bits_per_char <= 128) {
+        dict_ = std::make_unique<Dictionary<Kmer128>>();
+    } else if (odd_k * IDictionary::bits_per_char <= 256) {
+        dict_ = std::make_unique<Dictionary<Kmer256>>();
+    } else {
+        common::logger->error("k too big: {} > {}", odd_k, 256 / IDictionary::bits_per_char);
+        throw std::runtime_error("k fail");
+    }
+}
+
+DBGSSHash::DBGSSHash(std::string const& input_filename, size_t k, Mode mode)
+      : DBGSSHash(k, mode) {
+    sshash::build_configuration build_config;
+    build_config.k = k;
+    // quick fix for value of m... k/2 but odd
+    build_config.m = std::min(uint64_t((k_ + 1) / 2), sshash::constants::max_m);
+    build_config.verbose = common::get_verbose();
+    build_config.num_threads = get_num_threads();
+    if (build_config.m % 2 == 0)
+        build_config.m++;
+
+    // silence sshash construction messages when not verbose
+    if (!common::get_verbose())
+        std::cout.setstate(std::ios_base::failbit);
+
+    dict_->build(input_filename, build_config);
+    if (!common::get_verbose())
+        std::cout.clear();
+
+    num_nodes_ = dict_->size();
+}
+
+std::string DBGSSHash::file_extension() const {
+    return kExtension;
+}
+
+void DBGSSHash::add_sequence(std::string_view sequence,
+                             const std::function<void(node_index)>& on_insertion) {
+    throw std::runtime_error("adding sequences not implemented");
```

Comment:
nit: it seems that `std::logic_error` is considered more appropriate for these things. Also it should probably be "not supported" rather than "not implemented" because by design sshash is a static structure?

P.S. Though I suppose there are certain tricks that would make it expandable at the cost of extra $\log_2 n$ factor, where $n$ is the number of sequences...